### PR TITLE
refactor(core): cleanup feature flags for js source inclusion

### PR DESCRIPTION
**Pull request by nayeemrmn**
Remove `ExtensionFileSourceCode::LoadedFromFsDuringSnapshot` and feature `include_js_for_snapshotting` since they leak paths that are only applicable in this repo to embedders. Replace with feature `exclude_js_sources`. Additionally the feature `force_include_js_sources` allows negating it, if both features are set. We need both of these because features are additive and there must be a way of force including sources for snapshot creation while still having the `exclude_js_sources` feature. `force_include_js_sources` is only set for build deps, so sources are still excluded from the final binary.

You can also specify `force_include_js_sources` on any extension to override the above features for that extension. Towards #19398.

But there was still the snapshot-from-snapshot situation where code could be executed twice, I addressed that by making `mod_evaluate()` and scripts like `core/01_core.js` behave idempotently. This allowed unifying `ext::init_ops()` and `ext::init_ops_and_esm()` into `ext::init()`.

### CODE REVIEW

File: core/inspector.rs [hunk 1]
```diff
@@ -365,6 +365,9 @@ impl JsRuntimeInspector {
             w.poll_state = PollState::Parked;
             w.parked_thread.replace(thread::current());
           }
+          PollState::Parked => {
+            w.parked_thread.replace(thread::current());
+          }
```
Comment (bartlomieju) [inline core/inspector.rs:370]: Why was this change needed?

File: core/modules.rs [hunk 2]
```diff
@@ -459,29 +459,6 @@ impl ModuleLoader for ExtModuleLoader {
...
-          .to_string();
-      for m in unused_modules {
-        msg.push_str("  - ");
-        msg.push_str(m.0);
-        msg.push('\n');
-      }
-      panic!("{}", msg);
-    }
-  }
-}
```
Comment (bartlomieju) [inline core/modules.rs:483]: I don't understand why this was removed - it was very handy to assert we don't have files that were listed for inclusion in the snapshot but were never used. Could you explain it a bit @nayeemrmn?

File: Cargo.lock [hunk 3]
```diff
@@ -949,6 +949,7 @@ dependencies = [
  "anyhow",
  "bytes",
  "deno_ast",
+ "deno_core",
```
Comment (bartlomieju) [inline Cargo.lock:952]: I wonder if this will cause problems during publishing... I never seen crate depend on itself

File: core/extensions.rs [hunk 4]
```diff
@@ -608,54 +566,49 @@ impl ExtensionBuilder {
...
-  ($name:ident $($file:literal,)+) => {
+  ($name:ident $($file:literal,)*) => {
     vec![
       $($crate::ExtensionFileSource {
         specifier: concat!("ext:", stringify!($name), "/", $file),
-        code: $crate::ExtensionFileSourceCode::LoadedFromFsDuringSnapshot(
-          std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR")).join($file)
-        ),
-      },)+
+        code: include_str!($file),
```
Comment (bartlomieju) [inline core/extensions.rs:610]: This will cause the source code to be included in the resulting binary even though it's snapshotted. Which is visible in the benchmark page https://deno.com/benchmarks for binary size

File: core/runtime.rs [hunk 5]
```diff
@@ -1858,6 +1849,19 @@ impl JsRuntime {
...
+      return receiver;
+    }
+    if status == v8::ModuleStatus::Errored {
+      let (sender, receiver) = oneshot::channel();
+      let exception = module.get_exception();
+      sender
+        .send(exception_to_err_result(tc_scope, exception, false))
+        .unwrap();
+      return receiver;
+    }
```
Comment (bartlomieju) [inline core/runtime.rs:1864]: I really don't like this bit - while it might be fine for our own internal code, this will hide problems in module loading logic for user code

**Review by bartlomieju — COMMENTED** [review]
I'm late to the party, but I have a few questions. Overall I will revert this PR because it causes 3.5Mb regression in the binary size, because the source code is included in the binary for snapshotted code.

**Comment by bartlomieju** [comment]
Overall I like this cleanup - it would be ideal to have only a single method for defining extensions, but we need stricter control using cargo flags to decide if something should be included or not. I'm working on #19399 which also touches this subject, I will raise an alternative PR (which will probably build on this PR).

File: core/extensions.rs (hunk 4 above)
Comment (nayeemrmn) [reply core/extensions.rs:610]: It's supposed to work because no `ExtensionFileSource`s are registered in the final binary in the first place, they are replaced by an empty vec in `include_js_files!()`. I debug printed the esm files for extension builders after the `ext.esm()` call and they were all empty. But something apparently went wrong...

One difference this PR makes is that the static sources are included in the compilation of the `build.rs` scripts, where as before they were loaded at runtime of the `build.rs`. That shouldn't make a difference should it?

File: Cargo.lock (hunk 3 above)
Comment (nayeemrmn) [reply Cargo.lock:952]: It's a dev dependency to turn the force-include feature on for testing.

File: core/extensions.rs (hunk 4 above)
Comment (bartlomieju) [reply core/extensions.rs:610]: > One difference this PR makes is that the static sources are included in the compilation of the build.rs scripts, where as before they were loaded at runtime of the build.rs. That shouldn't make a difference should it?

I would assume that would be the case, I can't say for sure off the top of my head.

File: core/modules.rs (hunk 2 above)
Comment (nayeemrmn) [reply core/modules.rs:483]: The method didn't work anymore because there were now some redundant inclusions in `cli/build.rs`, and it seemed like we have a similar check at https://github.com/denoland/deno/blob/82dd90f98d6817caadcb9b6bd81aba80b2ee848d/core/runtime.rs#L895

File: core/inspector.rs (hunk 1 above)
Comment (nayeemrmn) [reply core/inspector.rs:370]: I meant to inquire if that was correct but forgot.. There were some inspector test failures on https://github.com/denoland/deno/pull/19463/commits/fc0f65a48307c2632fa23669be93f60f94caabdd because that variant was being hit now for some reason. That fixed it. But I don't know what changed there.

Comment (bartlomieju) [reply core/inspector.rs:370]: Huh, interesting! I would be interested in fixing this for sure, any chance you could try to create a test that would hit this code path? Inspector is quite brittle and esoteric bit of code so I would encourage a lot of caution when changing its logic.

Comment (nayeemrmn) [reply core/inspector.rs:370]: > any chance you could try to create a test that would hit this code path?

It's hit by the test failures on https://github.com/denoland/deno/actions/runs/5242046717/jobs/9464946684, caused by something in the rest of this PR.

File: core/modules.rs (hunk 2 above)
Comment (bartlomieju) [reply core/modules.rs:483]: Okay, I guess that makes sense.

**Comment by bartlomieju** [comment]
@nayeemrmn I suggest we land this PR in more bite-sized pieces - I like the idea of having only `ext::init()` and what is does is controlled via Cargo flags. Can you revert changes to `ExtensionFileSourceCode::LoadedFromFsDuringSnapshot` so we don't regress on the binary size?
